### PR TITLE
Add a struct version of types::Rectangle

### PR DESCRIPTION
**Pull request by dcampbell24**
Suggested way that Rectangle could be made a struct taking into consideration the types in conrod and how rectangles are used there. The hope is to reuse the structure in conrod.

CC #882

### CODE REVIEW

**Comment by zummenix** [comment]
Looks good, I like it.
`Dimensions` and `Rectangle` names looks a bit long to me. Maybe alternative names would be `Dim` or `Size`, and `Rect`.

File: src/types_new.rs [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+    /// Compute whether or not the point is inside the rectangle.
+    #[inline]
+    pub fn contains(&self, point: Point) -> bool {
+        self.left() < point.x && point.x < self.right() &&
+        self.top() < point.y && point.y < self.bottom()
+    }
+
+    /// Converts a rectangle into an array.
+    pub fn into_vec(self) -> [Scalar; 4] {
```
Comment (bvssvni) [inline src/types_new.rs]: Perhaps `into_array`?

**Comment by bvssvni** [comment]
I think this is on the right track! `Dim/Size` and `Rect` would be nice. What do you think about `Dim`?

**Comment by dcampbell24** [comment]
Rect is fine with me; although, I don't know why the type was changed from Rect to Rectangle previously.

Regarding Dim/Size I prefer Size since it is a whole word, but Size is also already defined in piston::window (in the same way minus the impl). Would there be any advantage to putting this in another crate?

**Comment by bvssvni** [comment]
@dcampbell24 It was not changed, but there was two types in the old design. `internal::Rectangle` was used as a type alias for consistency. `Rect` was a property type used in the piston-quack design. Now that we got a new design, we only need one.

**Comment by dcampbell24** [comment]
@bvssvni thanks for the clarification.

**Comment by bvssvni** [comment]
I think the members should be `pub`. These types are very simple to accessing members directly should be fine.

**Comment by zummenix** [comment]
- Do we need `Square` type?
Maybe add method to `Rect`: `pub fn square(pos: Point, len: Scalar) -> Rect;`

- Is it ok to rename `width` -> `w`, `height` -> `h`?

**Comment by bvssvni** [comment]
`Square` or `Circle` offer no performance advantage over `Rect` and `Ellipse`.

**Comment by bvssvni** [comment]
I think it is better to add a constructor for these instead of separate data types.

**Comment by bvssvni** [comment]
`w` and `h` instead of `width` and `height` is nice.

**Comment by dcampbell24** [comment]
Is from_circle and from_square OK? I don't like a function named "circle" because it sounds like it returns a circle when really it returns a rectangle that circumscribes the circle.

**Comment by bvssvni** [comment]
This library uses `new_` a lot. Perhaps `new_circle` and `new_square`?

Maybe we should rename to everything to `from_` but that's another discussion.

**Comment by dcampbell24** [comment]
My understanding is that `from` should be preferred over `new` when the constructor is a conversion constructor. Are the other uses of `new` in the library for doing conversions or just creating something with specific properties (which maybe should be called `with_property`)?

Is the square constructor creating a rectangle from a square or a rectangle that is a square? Is the circle constructor creating a rectangle from a circle or a rectangle that is a circle?

I am not sure what the answer is, but don't feel all that strongly either way, so am happy to go with `new_shape`. 

I guess another question for  a different issue then is also are we trying to adhere to the [Rust Guidelines](https://github.com/rust-lang/rust-guidelines).

References:
* http://aturon.github.io/ownership/constructors.html
* http://aturon.github.io/style/naming/README.html

**Comment by zummenix** [comment]
I propose to add following methods to `Rect`:
```rust
pub fn position(p: Point) -> Rect;
pub fn side(mut self, len: Scalar) -> Self;
pub fn get_center(&self) -> Point;
```
```rust
let square = Rect::position(...).side(...);
```
From this we can get a square rect, a center point for circle or ellipse and radius of a circle.

**Comment by zummenix** [comment]
We can also add `size` method similar to `side`.

**Comment by zummenix** [comment]
Alternative names for methods might be: `set_side`, `set_size`.

**Comment by bvssvni** [comment]
`set_size` should take `&mut self`, while size should take `self`.

**Comment by zummenix** [comment]
This is ready to merge.
@bvssvni What do you think?

**Comment by bvssvni** [comment]
Looks good!

**Pull request merged**